pkg/operator/controllers/cloudproviderconfig/cloudproviderconfig_controller.go

```diff
@@ -0,0 +1,100 @@
+package cloudproviderconfig
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers/base"
+	"github.com/Azure/ARO-RP/pkg/util/jsonutils"
+)
+
+const (
+	ControllerName    = "AROCloudProviderConfig"
+	controllerEnabled = "aro.cloudproviderconfig.enabled"
+)
+
+var cloudProviderConfigName = types.NamespacedName{Name: "cloud-provider-config", Namespace: "openshift-config"}
+
+// CloudProviderConfigReconciler reconciles the openshift-config/cloud-provider-config ConfigMap
+type CloudProviderConfigReconciler struct {
+	base.AROController
+}
+
+func NewReconciler(Log *logrus.Entry, client client.Client) *CloudProviderConfigReconciler {
+	return &CloudProviderConfigReconciler{
+		AROController: base.AROController{
+			Log:    Log,
+			Client: client,
+			Name:   ControllerName,
+		},
+	}
+}
+
+// Reconcile makes sure that the cloud-provider-config is healthy
+func (r *CloudProviderConfigReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance := &arov1alpha1.Cluster{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: arov1alpha1.SingletonClusterName}, instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(controllerEnabled) {
+		r.Log.Debug("controller is disabled")
+		return reconcile.Result{}, nil
+	}
+
+	r.Log.Debug("running")
+	return reconcile.Result{}, r.updateCloudProviderConfig(ctx)
+}
+
+// SetupWithManager setup our manager
+func (r *CloudProviderConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	cloudProviderConfigPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == cloudProviderConfigName.Name && o.GetNamespace() == cloudProviderConfigName.Namespace
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ConfigMap{}, builder.WithPredicates(cloudProviderConfigPredicate)).
+		Named(ControllerName).
+		Complete(r)
+}
+
+func (r *CloudProviderConfigReconciler) updateCloudProviderConfig(ctx context.Context) error {
+	cm := &corev1.ConfigMap{}
+	err := r.Client.Get(ctx, cloudProviderConfigName, cm)
+	if err != nil {
+		return fmt.Errorf("the ConfigMap openshift-config/cloud-provider-config is missing")
+	}
+
+	jsonConfig, ok := cm.Data["config"]
+	if !ok {
+		return fmt.Errorf("field config in ConfigMap openshift-config/cloud-provider-config is missing")
+	}
+
+	updateMap := map[string]string{"disableOutboundSNAT": "true"}
```

Comment:
From my understanding this will be encoded to a json string type like this `\"disableOutboundSNAT\": \"true\",`?  The underlying config map uses a bool value for this field and when testing the field as a string the load balancing rules do not get updated like they would if this was a json bool.  Is there a way to encode this as a boolean or maybe use some struct from upstream so we can marshal/unmarshal this data?

Reply:
The marshalling and unmarshalling of the data is happening in the utility function. This data is stored as a json string overall.

Reply:
Maybe I'm missing something but from my testing this has to be rendered as a boolean in the config map like `\"disableOutboundSNAT\": true`.  When I tested this as a string like `\"disableOutboundSNAT\": \"true\"` there was an issue with the updating the nodes and the nodes don't recover.  Maybe there is a struct upstream that we can use to marshal/unmarshal?

Reply:
The config map has the set up of:

```json
{ "config": { "disableOutboundSNAT": "true" } }
```

Reply:
FYI the upstream struct is missing this boolean setting.

https://github.com/openshift/cluster-api-provider-azure/blob/master/controllers/helpers.go#L318

Reply:
I did more research into this particular issue today. Even using a struct from the k8s upstream it does not unmarshal into a bool.

See: https://github.com/lranjbar/ARO-RP/pull/1

Reply:
Turns out we have a depreciated version of this struct in OpenShift in general. I had to paste in the struct from old upstream it get it work. All of the above conversation is moot. 🤦🏼‍♀️